pkg/lib/concurrency/async_result.go

```diff
@@ -0,0 +1,51 @@
+package concurrency
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// AsyncResult holds a value and an error, useful for async operations
+type AsyncResult[T any] struct {
+	Value T     `json:"value"`
+	Err   error `json:"error,omitempty"`
+}
+
+// MarshalJSON customizes the JSON representation of AsyncResult
+func (ar AsyncResult[T]) MarshalJSON() ([]byte, error) {
+	type Alias AsyncResult[T]
+	return json.Marshal(&struct {
+		Error string `json:"error,omitempty"`
+		*Alias
+	}{
+		Error: errorToString(ar.Err),
```

Comment:
could also do:
```suggestion
                Error: fmt.Sprintf("%s",ar.Err)
```
instead of a new method.

Reply:
not efficient, specially that this is going to be called each time a log record is serialized and deserialized when sent over the wire. A new method might be justified here